File: 1-EASY/Python/4-factorial-calculator/proposed-solution/main.py
# Challenge: Develop a recursive or iterative function to calculate the factorial of a non-negative integer.
# Consider handling special cases.

def factorial_recursive(n):
    if n < 0:
        raise ValueError("Factorial is not defined for negative numbers")
    if n == 0 or n == 1:
        return 1
    return n * factorial_recursive(n - 1)

def factorial_iterative(n):
    if n < 0:
        raise ValueError("Factorial is not defined for negative numbers")
    result = 1
    for i in range(2, n + 1):
        result *= i
    return result

# Example usage
if __name__ == "__main__":
    test_values = [0, 1, 5, 7]
    for val in test_values:
        print(f"Factorial of {val} (recursive): {factorial_recursive(val)}")
        print(f"Factorial of {val} (iterative): {factorial_iterative(val)}")

File: 1-EASY/JavaScript/4-factorial-calculator/proposed-solution/main.js
// Challenge: Develop a recursive or iterative function to calculate the factorial of a non-negative integer.
// Consider handling special cases such as negative numbers.

// Recursive factorial function
function factorialRecursive(n) {
    if (n < 0) {
        throw new Error("Factorial is not defined for negative numbers");
    }
    if (n === 0 || n === 1) {
        return 1;
    }
    return n * factorialRecursive(n - 1);
}

// Iterative factorial function
function factorialIterative(n) {
    if (n < 0) {
        throw new Error("Factorial is not defined for negative numbers");
    }
    let result = 1;
    for (let i = 2; i <= n; i++) {
        result *= i;
    }
    return result;
}

// Example usage
const testValues = [0, 1, 5, 7];
testValues.forEach(val => {
    console.log(`Factorial of ${val} (recursive): ${factorialRecursive(val)}`);
    console.log(`Factorial of ${val} (iterative): ${factorialIterative(val)}`);
});
